pkg/services/dashboards/database/database.go

```diff
@@ -594,24 +594,32 @@ func (d *dashboardStore) deleteDashboard(cmd *dashboards.DeleteDashboardCommand,
 		return dashboards.ErrDashboardNotFound
 	}
 
-	deletes := []string{
-		"DELETE FROM dashboard_tag WHERE dashboard_id = ? ",
-		"DELETE FROM star WHERE dashboard_id = ? ",
-		"DELETE FROM dashboard WHERE id = ?",
-		"DELETE FROM playlist_item WHERE type = 'dashboard_by_id' AND value = ?",
-		"DELETE FROM dashboard_version WHERE dashboard_id = ?",
-		"DELETE FROM dashboard_provisioning WHERE dashboard_id = ?",
-		"DELETE FROM dashboard_acl WHERE dashboard_id = ?",
+	type statement struct {
+		SQL  string
+		args []any
+	}
+
+	sqlStatements := []statement{
+		{SQL: "DELETE FROM dashboard_tag WHERE dashboard_id = ? ", args: []any{dashboard.ID}},
+		{SQL: "DELETE FROM star WHERE dashboard_id = ? ", args: []any{dashboard.ID}},
+		{SQL: "DELETE FROM dashboard WHERE id = ?", args: []any{dashboard.ID}},
+		{SQL: "DELETE FROM playlist_item WHERE type = 'dashboard_by_id' AND value = ?", args: []any{dashboard.ID}},
+		{SQL: "DELETE FROM dashboard_version WHERE dashboard_id = ?", args: []any{dashboard.ID}},
+		{SQL: "DELETE FROM dashboard_provisioning WHERE dashboard_id = ?", args: []any{dashboard.ID}},
+		{SQL: "DELETE FROM dashboard_acl WHERE dashboard_id = ?", args: []any{dashboard.ID}},
 	}
 
 	if dashboard.IsFolder {
 		// if this is a soft delete, we need to skip children deletion.
-		if !d.features.IsEnabledGlobally(featuremgmt.FlagDashboardRestore) && !cmd.SkipSoftDeletedDashboards {
-			deletes = append(deletes, "DELETE FROM dashboard WHERE folder_id = ? AND deleted IS NULL")
+		if !d.features.IsEnabledGlobally(featuremgmt.FlagDashboardRestore) {
+			sqlStatements = append(sqlStatements, statement{SQL: "DELETE FROM dashboard WHERE org_id = ? AND folder_uid = ? AND deleted IS NULL", args: []any{dashboard.OrgID, dashboard.UID}})
 
 			if err := d.deleteChildrenDashboardAssociations(sess, &dashboard); err != nil {
 				return err
 			}
+		} else {
+			// soft delete all dashboards in the folder
+			sqlStatements = append(sqlStatements, statement{SQL: "UPDATE dashboard SET deleted = ? WHERE org_id = ? AND folder_uid = ?", args: []any{time.Now(), dashboard.OrgID, dashboard.UID}})
```

Comment:
I like this approach but we are updating this row twice. We should remove the soft deletion from the [legacy_delete](https://github.com/grafana/grafana/blob/80a0de511f4cd234c9ec091104d974b818760d49/pkg/services/folder/folderimpl/folder.go#L855) function.

Reply:
I realised that this:
https://github.com/grafana/grafana/blob/9e31c82860be42517510e2b6ad3e40ceafee2e78/pkg/services/dashboards/database/database.go#L615
anyway affects only not soft deleted dashboards Ibecause of `AND deleted IS NULL`)
so I did the other way around after all; kept the `legacyDelete()` change and removed the checked:
could you check once again 🙏 ?

Reply:
The problem with this approach is the `deleteChildrenDashboardAssociations` in the next line will delete children data for soft deleted dashboards that could be restored in the future. The restoring won't be complete and certain [properties](https://github.com/grafana/grafana/blob/9e31c82860be42517510e2b6ad3e40ceafee2e78/pkg/services/dashboards/database/database.go#L693) will be lost.